mono/utils/mono-merp.c

```diff
@@ -0,0 +1,490 @@
+ *
+ * Author:
+ *   Alexander Kyte (alkyte@microsoft.com)
+ *
+ * (C) 2018 Microsoft, Inc.
+ *
+ */
+
+#include <config.h>
+#include <glib.h>
+
+#ifdef HOST_DARWIN
+#include "mono-merp.h"
+
+#include <unistd.h>
+#include <spawn.h>
+
+// OSX OS stuff now, for merpGUI interop
+#include <mach/mach.h>
+#include <mach/task_info.h>
+#include <mach/mach_types.h>
+#include <mach/mach_traps.h>
+#include <servers/bootstrap.h>
+
+typedef enum {
+	MerpArchInvalid = 0,
+
+	MerpArchx86_64 = 1,
+	MerpArchx86 = 2,
+	MerpArchPPC = 3,
+	MerpArchPPC64 = 4
+} MerpArch;
+
+typedef enum
+{
+	MERP_EXC_NONE = 0,
+
+	MERP_EXC_FORCE_QUIT = 1,
+	MERP_EXC_SIGSEGV = 2,
+	MERP_EXC_SIGABRT = 3,
+	MERP_EXC_SIGSYS  = 4,
+	MERP_EXC_SIGILL = 5,
+	MERP_EXC_SIGBUS = 6,
+	MERP_EXC_SIGFPE = 7 ,
+	MERP_EXC_SIGTRAP = 8,
+	MERP_EXC_SIGKILL = 9,
+	MERP_EXC_HANG  = 10
+} MERPExcType;
+
+typedef struct {
+	pid_t pidArg; // Process ID of crashed app (required for crash log generation)
+	MerpArch archArg; // Arch, MacOS only, bails out if not found also required for bucketization. (required)
+	uintptr_t capabilitiesArg; // App capabilities (optional) i.e. recover files, etc.
+	uintptr_t threadArg; // Stack Pointer of crashing thread. (required for crash log generation) Used for identifying crashing thread on crawl back when generating crashreport.txt
+	uintptr_t timeArg; // Total runtime (optional)
+
+	MERPExcType exceptionArg; // Exception type (refer to merpcommon.h and mach/exception_types.h for more info (optional)
+
+	const char *bundleIDArg; // App Bundle ID (required for bucketization)
+	const char *signatureArg; // App Bundle Signature (required) 
+
+	const char *versionArg; // App Version (required for bucketization)
+	const char *devRegionArg; // App region (optional)
+	const char *uiLidArg; // Application LCID aka Language ID (optional for bucketization)
+
+	const char *serviceNameArg; // This is the Bootstrap service name that MERP GUI will create to receive mach_task_self on a port created. Bails out if MERP GUI fails to receive mach_task_self from the crashed app. (Required for crash log generation)
+
+	const char *appLoggerName; // App loger name (optional)
+	const char *appLogSessionUUID; // App Session ID (optional but very useful)
+	gboolean isOfficeApplication; // Is office application (1 = True 0 = false). Needs to be 1 to continue. (Required by design)
+	gboolean isObjCException; // Is objectiveC Exception. (optional for crash log generation)
+	size_t memVirt; // Virtual memory (pagination) at the time of crash (optional)
+	size_t memRes; // Physical memory allocated at the time of the crash (optional)
+	const char *treSessionID; // Rules engine session ID (optional)
+	const char *exceptionCodeArg; // Exception code (optional)
+	const char *exceptionAddressArg; // Exception address (optional)
+	gboolean isAppRegisteredWithMAU; // Was this app installed with Microsoft Autoupdate (1 = True 0 = false)
+} MERPStruct;
```

Comment:
is there a public header where this is defined? otherwise, where is this information coming from?

Reply:
This implements an ad-hoc API that exists only as a set of fprintfs somewhere in a proprietary MS C file